lib/pbench/server/api/resources/pbench_users.py

```diff
@@ -0,0 +1,440 @@
+            abort(500, message="INTERNAL ERROR")
+
+        if not user:
+            self.logger.warning(
+                "No user found in the db for Username: {} while login", username
+            )
+            abort(403, message="No such user, please register first")
+
+        # Validate the password
+        if not check_password_hash(user.password, password):
+            self.logger.warning("Wrong password for user: {} while login", username)
+            abort(401, message="Wrong password")
+
+        try:
+            auth_token = self.auth.encode_auth_token(
+                self.token_expire_duration, user.id
+            )
+        except (
+            jwt.InvalidIssuer,
+            jwt.InvalidIssuedAtError,
+            jwt.InvalidAlgorithmError,
+            jwt.PyJWTError,
+        ):
+            self.logger.error(
+                "Could not encode the JWT auth token for user: {} while login", username
+            )
+            abort(
+                500, message="INTERNAL ERROR",
+            )
+
+        # check if the user is already logged in
+        # We would still proceed to login and return a new auth token to the user
+        if (
+            self.auth.token_auth.current_user()
+            and user == self.auth.token_auth.current_user()
+        ):
+            self.logger.warning("User already logged in and trying to re-login")
+            if self.auth.token_auth.get_auth()["token"] == auth_token.decode():
+                # If a user attempts to re-login immediately within a second,
+                # the new auth token will be same as the one present in the header
+                # Therefore we will not fulfill the re-login request
+                self.logger.warning("Attempted immediate re-login")
+                abort(403, message="Retry login after some time")
+
+        # Add the new auth token to the database for later access
+        try:
+            token = ActiveTokens(token=auth_token.decode())
+            # TODO: Decide on the auth token limit per user
+            user.auth_tokens.append(token)
+
+            # insert the auth token
+            Database.db_session.add(token)
+            Database.db_session.commit()
+            self.logger.info("New auth token registered for user {}", user.email)
+        except IntegrityError:
+            self.logger.warning(
+                "Duplicate auth token got created, user might have tried to re-login immediately"
+            )
+            Database.db_session.rollback()
+            abort(409, message="Retry login after some time")
+        except SQLAlchemyError as e:
+            self.logger.error(
+                "SQLAlchemy Exception while logging in a user {}", type(e)
+            )
+            Database.db_session.rollback()
+            abort(500, message="INTERNAL ERROR")
+        except Exception:
+            self.logger.exception("Exception while logging in a user")
+            Database.db_session.rollback()
```

Comment:
I don't like the necessity of scattering the rollbacks around the caller; they should be encapsulated. My encapsulated wrappers internalize that to hide it; I think this may be the best argument for such wrappers. E.g.,:

```
    def add(self):
        """
        add Add the Dataset object to the database
        """
        try:
            Database.db_session.add(self)
            Database.db_session.commit()
        except Exception as e:
            Database.db_session.rollback()
            raise DatasetSqlError(f"Error adding {self}: {e}")
```